homeassistant/components/websocket_api/http.py

```diff
@@ -100,46 +98,53 @@ async def _writer(self) -> None:
+            while not self.wsock.closed:
+                if (messages_remaining := len(message_queue)) == 0:
+                    self._ready_future = loop.create_future()
+                    await self._ready_future
+                    messages_remaining = len(message_queue)
+
+                # A None message is used to signal the end of the connection
+                if (process := message_queue.popleft()) is None:
+                    return
+
+                messages_remaining -= 1
+                message = process if isinstance(process, str) else process()
+
+                if (
+                    not messages_remaining
+                    or not self.connection
+                    or not self.connection.can_coalesce
+                ):
+                    if is_enabled_for(logging_debug):
+                        debug("%s: Sending %s", self.description, message)
+                    await send_str(message)
+                    continue
 
+                messages: list[str] = [message]
+                while messages_remaining:
                     # A None message is used to signal the end of the connection
                     if (process := message_queue.popleft()) is None:
                         return
-
+                    messages.append(process if isinstance(process, str) else process())
                     messages_remaining -= 1
-                    message = process if isinstance(process, str) else process()
-
-                    if (
-                        not messages_remaining
-                        or not self.connection
-                        or not self.connection.can_coalesce
-                    ):
-                        debug("Sending %s", message)
-                        await send_str(message)
-                        continue
-
-                    messages: list[str] = [message]
-                    while messages_remaining:
-                        # A None message is used to signal the end of the connection
-                        if (process := message_queue.popleft()) is None:
-                            return
-                        messages.append(
-                            process if isinstance(process, str) else process()
-                        )
-                        messages_remaining -= 1
-
-                    joined_messages = ",".join(messages)
-                    coalesced_messages = f"[{joined_messages}]"
-                    debug("Sending %s", coalesced_messages)
-                    await send_str(coalesced_messages)
+
+                joined_messages = ",".join(messages)
+                coalesced_messages = f"[{joined_messages}]"
+                if is_enabled_for(logging_debug):
+                    debug("%s: Sending %s", self.description, coalesced_messages)
+                await send_str(coalesced_messages)
+        except asyncio.CancelledError:
+            debug("%s: Writer cancelled", self.description)
```

Comment:
What with futures.CancelledError?

Reply:
I’m pretty sure that’s left over from a previous iteration where there was executor usage.  I don’t see any place where this can happen in the current design

Reply:
The linked issue showed executor used inside aiohttp. But maybe that is gone now.

Reply:
To be safe we should probably trap it and reraise as asyncio.CanceledError

Reply:
After digging some more into aiohttp that code no longer exists so the problem isn't possible to happen anymore

https://github.com/home-assistant/core/issues/9546

https://github.com/aio-libs/aiohttp/blame/master/aiohttp/base_protocol.py#L81

Reply:
removed in 09839c955f1489e2ce715294af6c06aa1e848fb7